exec/memory/base/src/main/java/org/apache/drill/exec/memory/AllocationManager.java

```diff
@@ -253,10 +261,12 @@ public boolean transferBalance(final BufferLedger target) {
           target.historicalLog.recordEvent("incoming(from %s)", owningLedger.allocator.name);
         }
 
-        boolean overlimit = target.allocator.forceAllocate(size);
+        // Release first to handle the case where the current and target allocators were part of the same
+        // parent / child tree.
         allocator.releaseBytes(size);
+        boolean allocationFit = target.allocator.forceAllocate(size);
```

Comment:
move `owningLedger = target;` before `return target.allocator.forceAllocate(size)` as there is no check for the result of the call anyway.

Reply:
If this happens, is not there a problem that the old allocator already released the memory? In any case, won't runtime exception cancel the query anyway and all allocators will be closed.

Reply:
What about debugging:
- An operation fails
- And yet there is a change of ownership
 
@vrozov, I rather not make this change as I strongly believe the change of ownership should happen only on success.

Reply:
In this case, changing the order of `forceAllocate()` and `releaseBytes()` is incorrect as ownership is not changed, but the old owner does not account for that memory anymore.

Reply:
- The change of order is an optimization for a parent / child relationship as if we don't release first, then we could unnecessarily go over the memory budget (double counting).
- The force-alloc() / free() failures should never happen on normal conditions; when they do, the best thing to do is to exit. I still prefer not to promote the target allocator till it is 100% successful.